packages/ethereum/contracts/src/Crypto.sol

```diff
@@ -0,0 +1,598 @@
+      // store data assembly-favouring ways
+      mstore(p, 0x20)             // Length of Base
+      mstore(add(p, 0x20), 0x20)  // Length of Exponent
+      mstore(add(p, 0x40), 0x20)  // Length of Modulus
+      mstore(add(p, 0x60), el)  // Base
+      mstore(add(p, 0x80), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2D) // p - 2
+      mstore(add(p, 0xa0), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F) // p
+      if iszero(staticcall(not(0), 0x05, p, 0xC0, p, 0x20)) {
+        revert(0, 0)
+      }
+
+      o := mload(p)
+    }
+  } 
+
+  /**
+   * Performs a modular exponentiation using the expmod precompile.
+   */
+  function expmod(uint256 base, uint256 exponent, uint256 modulus) public view returns (uint256 o) {
+    assembly {
+      //  define pointer
+      let p := mload(0x40)
+      // store data assembly-favouring ways
+      mstore(p, 0x20)             // Length of Base
+      mstore(add(p, 0x20), 0x20)  // Length of Exponent
+      mstore(add(p, 0x40), 0x20)  // Length of Modulus
+      mstore(add(p, 0x60), base)  // Base
+      mstore(add(p, 0x80), exponent)    // Exponent
+      mstore(add(p, 0xa0), modulus)     // Modulus
+      if iszero(staticcall(not(0), 0x05, p, 0xC0, p, 0x20)) {
+        revert(0, 0)
+      }
+
+      o := mload(p)
+    }
+  }
+
+  /**
+   * Converts a curve point to an Ethereum address.
+   *
+   * This function can be used to witness the result of a scalar
+   * multiplication.
+   */
+  function pointToAddress(CurvePoint calldata p) public pure returns (address) {
+    return address(uint160(uint256(keccak256(abi.encodePacked(p.x, p.y)))));
+  }
+
+  /**
+   * Adds two elliptic curve points using the general implementation.
+   *
+   * This function is optimized to perform one point addition.
+   *
+   * Optimizations:
+   * - solidity assembly
+   * - optimize for a single addition
+   */
+  function ecAdd(CurvePoint memory p, CurvePoint memory q) public view returns (CurvePoint memory r)  {
+    if (p.x == 0 && p.y == 0) { // Inf + Q = Q
+      return q;
+    }
+
+    if (q.x == 0 && q.y == 0) { // P + Inf = P
+      return p;
+    }
+
+    assembly {
+      let toInvert := addmod( // q.x - p.x
+        mload(q), // q.x 
+        sub(SECP256K1_BASE_FIELD_ORDER, mload(p)), // - p.x
+        SECP256K1_BASE_FIELD_ORDER
+      )
+
+      let payload := mload(0x40)
+      mstore(payload, 0x20)             // Length of Base
+      mstore(add(payload, 0x20), 0x20)  // Length of Exponent
+      mstore(add(payload, 0x40), 0x20)  // Length of Modulus
+      mstore(add(payload, 0x60), toInvert)  // Base
+      mstore(add(payload, 0x80), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2D) // p - 1
+      mstore(add(payload, 0xa0), SECP256K1_BASE_FIELD_ORDER)     // Modulus
+      if iszero(staticcall(not(0), 0x05, payload, 0xC0, payload, 0x20)) {
+        revert(0, 0)
+      }
+
+      let lambda := mulmod( // (q.y - p.y) * (q.x - p.x) ^ -1
```

Comment:
This formula does not work if P = Q. 
You need to detect point doubling or use some complete formulas for secp256k1 if they exist.

Reply:
also does not work if Q = -P